ipn/ipnlocal/local.go

```diff
@@ -1557,10 +1557,13 @@ func endpointsEqual(x, y []tailcfg.Endpoint) bool {
 	return true
 }
 
-func (b *LocalBackend) SetNotifyCallback(notify func(ipn.Notify)) {
+// SetNotifyCallbackForTest sets the function to call when the backend has
+// something to notify the frontend about. It must be called before Start and is
+// only for use by tests.
+func (b *LocalBackend) SetNotifyCallbackForTest(notify func(ipn.Notify)) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	b.notify = notify
```

Comment:
So you even need the lock unlock here now?

Reply:
it changed a bit, ptal.

Reply:
I had to migrated the func watcher to run a goroutine instead as otherwise i'd get into situations where the notify callback (in other repos) would try to call back into LocalBackend and try to acquire a lock and would deadlock.